### ISSUE TITLE
Babel.js for PHP — Transpile the *other* way

### ISSUE THREAD

**Issue by skyzyx** [rectorphp/rector#7460]
# Feature Request

As a long-time vendor of some popular open-source software, I've wanted for a long time something like Babel.js for PHP. Something where I can write code for PHP 8.x, then _transpile_ it back to older versions of PHP that my customers run.

This appears to be doing the exact opposite. :) But, the logic here to move a project in one direction is very similar to the logic required to move in the other direction.

So that's the request — a project that allows me to write clean, modern code, then transpile it back to older versions of PHP so that customers can run it in their — likely older — environments.

(I'm not saying that running EOL versions of PHP is in any way a good idea. But reality dictates that customers like this exist, and I've love to support them without hating life.)

**Comment by TomasVotruba** [comment]
Hi, that's actually the way Rector is build on PHP 8.1 and released on PHP 7.2 :)

See:

* https://github.com/rectorphp/rector-downgrade-php
* https://getrector.org/blog/how-all-frameworks-can-bump-to-php-81-and-you-can-use-older-php
* https://tomasvotruba.com/blog/how-to-develop-sole-package-in-php81-and-downgrade-to-php72/
* https://blog.logrocket.com/transpiling-php-code-from-8-0-to-7-x-via-rector/
* https://blog.logrocket.com/phabel-vs-rector-better-transpiling-php-code/

**Comment by skyzyx** [comment]
I just heard about this project last week, so I must've missed that in the docs. But I'm really excited that this project exists.

Thanks!

**Comment by TomasVotruba** [comment]
Welcome to the laziest PHP devs club :+1: 

Btw, I'm curious, what do you want to use the downgrade for? :)

**Comment by skyzyx** [comment]
I created a project called SimplePie back in 2004. It's — I think — the world's most-used RSS feed parser.

As I've grown as a software engineer over the years and adopted modern features, most PHP devs have not. (This is proven by WordPress still supporting 5.6.) In a new SimplePie-related project I'm working on, I want to be able to use the latest features (e.g., 8.2), but still support all other non-EOL version of PHP (e.g., 7.4, 8.0, 8.1). Yes, I'll be providing optimized Docker images, but a large number of PHP projects still haven't taken the (obvious, IMO) step of adopting Docker yet.

Babel.js is great at solving this problem in JavaScript. It understands the language well enough to be able to translate things in newer ECMAScript editions back to older versions shipping in browsers. I've wanted something like this for PHP for a long time. I know that `@nikic` has had the PHP Parser project for a while, and I've dabbled in building something, but just couldn't justify the time required to see it though.

I'm thrilled this exists.
